extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,335 @@
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    /** @private {!Promise} */
+    this.activatedPromise_ = null;
+
+    assert(this.element.hasAttribute('id'), 'amp-user-notification should ' +
+        'have an id.');
+
+    /** @private {string} */
+    this.elementId_ = this.element.getAttribute('id');
+
+    /** @private {?string} */
+    this.ampUserId_ = null;
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        'amp-user-notification should have `data-show-if-href` attribute.');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    assert(this.element.hasAttribute('data-dismiss-href'),
+        'amp-user-notification should have `data-dismiss-href` attribute.');
+
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    this.options_ = {
+      credentials: this.element.getAttribute('credentials')
+    };
+
+    /** @private {?function(*)} */
+    this.layedoutResolve_ = null;
+
+    /** @private {?function(*)} */
+    this.dialogResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.layedoutPromise_ = new Promise((resolve, reject) => {
+      this.layedoutResolve_ = resolve;
+    });
+
+    /** @private @const {!Promise} */
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.dialogResolve_ = resolve;
+    });
+
+    this.userNotification_.registerUserNotification(
+        this.elementId_, this.layedoutResolve_, this.dialogPromise_);
+
+    this.registerAction('dismiss', this.dismiss.bind(this));
+  }
+
+  /** @override */
+  layoutCallback(visible) {
+    if (!this.activatedPromise_) {
+    // `layedoutPromise_` will block until there is no other
+    // `amp-user-notification` in display at this point in time.
+    this.activatedPromise_ = this.layedoutPromise_
+        .then(this.getAsyncCid_.bind(this))
+        .then(this.fetchShowEndpoint_.bind(this))
+        .then(this.onFetchShowEndpointSuccess_.bind(this))
+        .catch(this.onRejection_.bind(this));
+    }
+    return this.activatedPromise_;
+  }
+
+  /**
+   * Executes a `GET` request to the href given on the `data-show-if-href`
+   * attribute.
+   * @param {string} ampUserId
+   * @return {!Promise}
+   * @private
+   */
+  fetchShowEndpoint_(ampUserId) {
+    this.ampUserId_ = ampUserId;
+    const getReq = {
+      /** @type {GetRequestMetadata} */
+      body: {
+        elementId: this.elementId_,
+        ampUserId: ampUserId
+      }
+    };
+
+    return xhrFor(this.win_).fetchJson(this.showIfHref_, getReq);
+  }
+
+  /**
+   * Creates an POST to the specified `data-dismiss-href` url.
+   * @private
+   * @return {!Promise}
+   */
+  postDismissEnpoint_() {
+    return xhrFor(this.win_).fetchJson(this.dismissHref_, {
+      method: 'POST',
+      body: {
+        elementId: this.elementId_,
```

Comment:
just a way for the server to identify which notification component this came from.